### PR TITLE
Extract :buildOperations subproject

### LINKED ISSUE #13934: Extract build operations to separate subproject
---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by lptr**
Fixes #13934 

- [x] Add new `:buildOperations` subproject
- [x] Move reusable parts of build operations to `:buildOperations`
- [x] Abstract out `BuildOperationCategory`
- [ ] Add test for `AbstractBuildOperationRunner` (extract common parts from `DefaultBuildOperationExecutorTest`)

## Open questions

- Is it useful to share the managed thread concept?

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle test sanity please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test RFM please

File: subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+            Thread.currentThread().interrupt();
+        }
+        if (t instanceof RuntimeException) {
+            throw (RuntimeException) t;
+        }
+        if (t instanceof Error) {
+            throw (Error) t;
+        }
+        if (t instanceof IOException) {
+            throw new RuntimeException(t.getMessage(), t);
```
Comment (wolfs) [inline subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java:234]: Why is here the extra branch for IOException? Deleting this `if` should yield the same result, right?

Comment (lptr) [reply subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java:234]: Ah, right. We used to convert this to `UncheckedIOException`, but that doesn't work with Java 6... :/

Comment (lptr) [reply subprojects/build-operations/src/main/java/org/gradle/internal/operations/DefaultBuildOperationRunner.java]: Fixed.

**Comment by lptr** [comment]
@bot-gradle test QFL please

File: subprojects/build-operations/build-operations.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    id("gradlebuild.publish-public-libraries")
+}
+
+description = "Build operations are our way to inspect the process of executing a build"
+
+gradlebuildJava.usedInWorkers()
+
+dependencies {
+    implementation(project(":baseAnnotations"))
+    implementation(libs.guava)
```
Comment (wolfs) [inline subprojects/build-operations/build-operations.gradle.kts:27]: Can we get rid of the `guava` dependency`?

**Comment by lptr** [comment]
@bot-gradle test QFL please

File: subprojects/build-operations/build-operations.gradle.kts (hunk 2 above)
Comment (lptr) [reply subprojects/build-operations/build-operations.gradle.kts]: I removed the Guava dependency, and also removed a use of a Java 8 API. Java 6 compatibility should be airtight now.

Comment (wolfs) [reply subprojects/build-operations/build-operations.gradle.kts]: Nice!

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
Alright, I'm going to merge as soon as the build passes. We can address feedback on the PR afterwards. I'd like to also switch to composition for the functional split between `BuildOperationRunner` and `BuildOperationExecutor` in a separate PR. Doing so will enable better names for the two types, and will also allow the tests to be split in a meaningful way.

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Comment by lptr** [comment]
@bot-gradle test QFL please

**Pull request merged**